### PR TITLE
feat: support w3c revocation

### PR DESCRIPTION
**Pull request by KulkarniShashank**
### What ###
This PR introduces support for W3C credential revocation. The following changes have been made:

Refactor functionality for Verify Credential and Verify Presentation.
Integrated with the Bit String Status List for checking and updating credential status.

### How ###
**Revocation Flow**: Implemented the revocation flow by checking a credential status and updating the credential issuance process to handle revocable credentials.
**Integration**: Integrated the new revocation features with the existing system, ensuring compatibility and proper handling of revoked credentials.

### CODE REVIEW

File: packages/core/src/modules/vc/data-integrity/W3cJsonLdCredentialService.ts [hunk 1]
```diff
@@ -265,6 +304,52 @@ export class W3cJsonLdCredentialService {
         challenge: options.challenge,
         domain: options.domain,
         documentLoader: this.w3cCredentialsModuleConfig.documentLoader(agentContext),
+        checkStatus: async ({ credential }: { credential: W3cJsonCredential }) => {
```
Comment (TimoGlastra) [inline packages/core/src/modules/vc/data-integrity/W3cJsonLdCredentialService.ts:307]: Can you extract this into a method and reuse that for both checkStatus methods?

File: packages/core/src/modules/vc/data-integrity/W3cJsonLdCredentialService.ts [hunk 2]
```diff
@@ -265,6 +304,52 @@ export class W3cJsonLdCredentialService {
...
+            const bitStringCredential =
+              (await bitStringStatusListCredential.json()) as unknown as BitStringStatusListCredential
+            const encodedBitString = bitStringCredential.credential.credentialSubject.encodedList
+            const compressedBuffer = Uint8Array.from(atob(encodedBitString), (c) => c.charCodeAt(0))
+
+            // Decompress using pako
+            const decodedBitString = pako.ungzip(compressedBuffer, { to: 'string' })
+            const statusListIndex = Number(credential.credentialStatus.statusListIndex)
+
+            if (statusListIndex < 0 || statusListIndex >= decodedBitString.length) {
```
Comment (TimoGlastra) [inline packages/core/src/modules/vc/data-integrity/W3cJsonLdCredentialService.ts:341]: We should validate the bitstring status list credential as well

File: packages/core/src/modules/vc/models/credential/W3cJsonCredential.ts [hunk 3]
```diff
@@ -10,4 +10,59 @@ export interface W3cJsonCredential {
   expirationDate?: string
   credentialSubject: SingleOrArray<JsonObject>
   [key: string]: unknown
+  credentialStatus?: SingleOrArray<CredentialStatus>
+}
+
+type CredentialStatusType = 'BitstringStatusListEntry'
```
Comment (TimoGlastra) [inline packages/core/src/modules/vc/models/credential/W3cJsonCredential.ts:16]: Can we move all the status list logic related to bit string status list to a separate file?

File: packages/core/src/modules/vc/data-integrity/W3cJsonLdCredentialService.ts [hunk 4]
```diff
@@ -265,6 +304,52 @@ export class W3cJsonLdCredentialService {
...
+                'Verifying credential status as an array for JSON-LD credentials is currently not supported'
+              )
+            }
+
+            // Ensure credentialStatus contains the necessary properties
+            if (!credential.credentialStatus || credential.credentialStatus.statusListIndex === undefined) {
+              throw new CredoError('Invalid credential status format')
+            }
+
+            const credentialStatusURL = credential.credentialStatus.statusListCredential
```
Comment (TimoGlastra) [inline packages/core/src/modules/vc/data-integrity/W3cJsonLdCredentialService.ts:321]: We should first check the type fiets to make sure this is a bit string status type. There could be others and those are not supported at the moment.

**Review by TimoGlastra — CHANGES_REQUESTED** [review]
Thanks for this! Left some comments

**Comment by KulkarniShashank** [comment]
@TimoGlastra As per your suggestions, I have made the changes. Please review them and let me know your thoughts.

File: packages/core/src/modules/vc/data-integrity/W3cJsonLdCredentialService.ts (hunk 1 above)
Comment (KulkarniShashank) [reply packages/core/src/modules/vc/data-integrity/W3cJsonLdCredentialService.ts:268]: @TimoGlastra I have developed a common function for this purpose. However, in the **checkStatus** method, there are specific conditions related to credential status that are dependent on the payload received from the **verifyCredential** and **verifyPresentation** functions. Therefore, the **checkStatus** method invokes the common function, as it contains the identical logic.

**Comment by Tweeddalex** [comment]
Out of interest @KulkarniShashank will this work with cheqd DIDs and on-ledger Bitstring Status Lists?

**Comment by KulkarniShashank** [comment]
> Out of interest @KulkarniShashank will this work with cheqd DIDs and on-ledger Bitstring Status Lists?

Its can work for all the DIDs that support JSON-LD credential format

**Comment by TimoGlastra** [comment]
However @Tweeddalex this PR does not support cheqd yet. The current implementation only suppors https and uses `fetch` to retrieve the status list. For cheqd support we'd have to fetch the status list differently.

File: packages/core/src/modules/vc/data-integrity/W3cJsonLdCredentialService.ts [hunk 5]
```diff
@@ -265,6 +265,14 @@ export class W3cJsonLdCredentialService {
         challenge: options.challenge,
         domain: options.domain,
         documentLoader: this.w3cCredentialsModuleConfig.documentLoader(agentContext),
+        checkStatus: async ({ credential }: { credential: W3cJsonCredential }) => {
+          if ('credentialStatus' in credential) {
+            await verifyBitStringCredentialStatus(credential, agentContext)
+          }
```
Comment (TimoGlastra) [inline packages/core/src/modules/vc/data-integrity/W3cJsonLdCredentialService.ts:271]: verifyPresentation also has options.verifyCredentialStatus, can you add that check here as well?

File: packages/core/src/modules/vc/models/credential/W3cCredentialStatus.ts [hunk 6]
```diff
@@ -1,23 +1,61 @@
...
+  @IsString()
+  public type!: CredentialStatusType
+
+  @IsEnum(CredentialStatusPurpose)
+  @IsString()
+  public statusPurpose!: CredentialStatusPurpose
+
   @IsString()
-  public type!: string
+  public statusListIndex!: string
```
Comment (TimoGlastra) [inline packages/core/src/modules/vc/models/credential/W3cCredentialStatus.ts:42]: Looking at the VC Data model specification, these properties are not required fro the `credentialStatus` property. Only `id` and `type` are. 

We should have a generic base class and allow multiple extension points. You can look at e.g. how we solved this with the DidDocument service property, where there's a base service class and multiple possible extensions that are matched based on the `type` property

File: packages/core/src/modules/vc/models/credential/W3cCredentialStatus.ts [hunk 7]
```diff
@@ -1,23 +1,61 @@
...
+  try {
+    await validateOrReject(entry)
+    return true
+  } catch (errors) {
+    agentContext.config.logger.debug(`Credential status validation failed: ${errors}`, {
+      stack: errors,
+    })
+    throw new CredoError(`Invalid credential status type: ${errors}`)
+  }
 }
```
Comment (TimoGlastra) [inline packages/core/src/modules/vc/models/credential/W3cCredentialStatus.ts:61]: This validation is automatically done when transforming into a specific class instance type, so I don't think it's needed if you follow the did document service approach

File: packages/core/src/modules/vc/models/credential/W3cJsonCredentialStatus.ts [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+// Define an interface for the `credential` object that uses `CredentialSubject`
+export interface Credential {
+  credentialSubject: CredentialSubject
+}
+
+// Use the `Credential` interface within `BitStringStatusListCredential`
+export interface BitStringStatusListCredential {
+  credential: Credential
+}
```
Comment (TimoGlastra) [inline packages/core/src/modules/vc/models/credential/W3cJsonCredentialStatus.ts:55]: All these interface names should be bitstring status list specific. It is important to make a distinction between bitstring status list, and possible other `credentialStatus` methods that will be added in the future.

File: packages/core/src/modules/vc/data-integrity/VerifyBitStringCredentialStatus.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  try {
+    return (await response.json()) as BitStringStatusListCredential
+  } catch (error) {
+    throw new CredoError('Failed to parse the bit string status list credential')
+  }
+}
+
+export const verifyBitStringCredentialStatus = async (credential: W3cJsonCredential, agentContext: AgentContext) => {
```
Comment (TimoGlastra) [inline packages/core/src/modules/vc/data-integrity/VerifyBitStringCredentialStatus.ts:28]: This doesn't validate the signature of the status list credential yet

File: packages/core/src/modules/credentials/CredentialsApi.ts [hunk 10]
```diff
@@ -515,6 +525,87 @@ export class CredentialsApi<CPs extends CredentialProtocol[]> implements Credent
...
 
+  /**
+   * Revoke a credential by issuer
+   * associated with the credential record.
+   *
+   * @param credentialRecordId The id of the credential record for which to revoke the credential
+   * @returns Revoke credential notification message
+   *
+   */
+  public async revokeJsonLdCredential(options: RevokeCredentialOption): Promise<{ message: string }> {
```
Comment (TimoGlastra) [inline packages/core/src/modules/credentials/CredentialsApi.ts:536]: there is a lot of dupcliation in this file with the already present methods. Some suggestions:
- move all the credential status logic into a specific folder in the w3c module
- add the revocation related methods to the w3c credentails api, not the DIDcomm credentials api
- the only thing we should add it here is a `sendRevocationNotification` method (which we can extend to support non-anoncreds revocadion notification messages as well)
- making a POST to the credential status list url to update the status list seems implementation specific and may not work for all implementations

**Comment by genaris** [comment]
Superseded by #2072

**Pull request closed without merging**